oak-core-spi/src/main/java/org/apache/jackrabbit/oak/osgi/OsgiWhiteboard.java

```
@@ -69,13 +69,13 @@ public <T> Registration register(
         Objects.requireNonNull(properties);
         checkArgument(type.isInstance(service));
 
-        Dictionary<Object, Object> dictionary = new Hashtable<Object, Object>();
+        Dictionary<String, Object> dictionary = new Hashtable<>();
         for (Map.Entry<?, ?> entry : properties.entrySet()) {
-            dictionary.put(entry.getKey(), entry.getValue());
+            dictionary.put((String)entry.getKey(), entry.getValue());
```

Is a cast safe here? Wouldn't be "toString()" better?

---

cast is safe, because `Property` always only expose String, String pairs (https://docs.oracle.com/javase/8/docs/api/java/util/Properties.html). Compare with https://stackoverflow.com/questions/873510/why-does-java-util-properties-implement-mapobject-object-and-not-mapstring-st

---

Actually the cast is no even necessary here, I am gonna remove it.

---

@reschke Is changing this SPI method signature ok to only accept Map<String, ?> because that is the only one OSGi accepts (see https://docs.osgi.org/javadoc/r4v43/core/org/osgi/framework/BundleContext.html#registerService(java.lang.Class,%20S,%20java.util.Dictionary) or do we need to maintain backwards compatibility here as well?

---

if we change it, BND will claim that's an incompatible change, no?

---

Not sure, as this as Generics a compile time thing only, but I left the signature as is and clarified the javadoc just a bit. But throwing a ClassCastException is better than silently using the toString() so I left the cast.